toolkit/tools/pkg/imagecustomizerlib/customizepackages.go

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package imagecustomizerlib
+
+import (
+	"fmt"
+	"path"
+	"strings"
+
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/imagecustomizerapi"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/logger"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/safechroot"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/shell"
+)
+
+func updatePackages(buildDir string, baseConfigPath string, packagesToAddLists []string, packagesToAdd []string,
+	packagesToRemoveLists []string, packagesToRemove []string, imageChroot *safechroot.Chroot, rpmsSources []string,
+	useBaseImageRpmRepos bool,
+) error {
+	var err error
+
+	allPackagesToRemove, err := collectPackagesList(baseConfigPath, packagesToRemoveLists, packagesToRemove)
+	if err != nil {
+		return err
+	}
+
+	allPackagesToAdd, err := collectPackagesList(baseConfigPath, packagesToAddLists, packagesToAdd)
+	if err != nil {
+		return err
+	}
+
+	err = removePackages(buildDir, allPackagesToRemove, imageChroot)
+	if err != nil {
+		return err
+	}
+
+	err = installPackages(buildDir, allPackagesToAdd, imageChroot, rpmsSources, useBaseImageRpmRepos)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func collectPackagesList(baseConfigPath string, packageLists []string, packages []string) ([]string, error) {
+	var err error
+
+	// Read in the packages from the package list files.
+	var allPackages []string
+	for _, packageListRelativePath := range packageLists {
+		packageListFilePath := path.Join(baseConfigPath, packageListRelativePath)
+
+		var packageList imagecustomizerapi.PackageList
+		err = imagecustomizerapi.UnmarshalYamlFile(packageListFilePath, &packageList)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read package list file (%s):\n%w", packageListFilePath, err)
+		}
+
+		allPackages = append(allPackages, packageList.Packages...)
+	}
+
+	allPackages = append(allPackages, packages...)
+	return allPackages, nil
+}
+
+func removePackages(buildDir string, allPackagesToRemove []string, imageChroot *safechroot.Chroot) error {
+	var err error
+
+	tnfRemoveArgs := []string{
+		"-v", "remove", "--assumeyes", "--disablerepo", "*",
```

Comment:
Maybe of interest is `autoremove`. It will pull any dependencies as well as the main package.
Warning: I've seen it pull packages unexpectedly sometimes that shouldn't get yanked... like `rpm` and `tdnf` itself...

Reply:
Hmm, I think `autoremove` could result in unexpected behavior for users since the packages that get removed could change when the package dependencies are updated. So, I think sticking with `remove` will be better. (Though perhaps we can mention in the docs that `autoremove` can be used to find additional packages that can removed.)